### ISSUE TITLE
Existential operator in new method of a class

### ISSUE THREAD

**Issue by Le0Developer** [pigpigyyy/Yuescript#55]
Hello, I'd like to request a feature that allows the use of the existential operator alongside the @ in the new method (especially useful for optional values that fallback to the class's value).

e.g.
```moon
class Person
    name: "Default Name"
    new: (@name?) =>
```
equivalent to
```moon
class Person
    name: "Default Name"
    new: (name) =>
        @name = name if name?
```

**Comment by CriztianiX** [comment]
Mmmm...
Why you do not use

```
class Person
    new: (name = 'Default Name' ) =>
        @name = name
```

**Comment by Le0Developer** [comment]
There's one major reason: Inheritance

```moon
class Person
    name: "Default Name"
    new: (@name?) =>

class AnoterPerson extend Person
    name: "Another Name"
```

**Comment by CriztianiX** [comment]
> There's one major reason: Inheritance
> 
> ```moonscript
> class Person
>     name: "Default Name"
>     new: (@name?) =>
> 
> class AnotherPerson extends Person
>     name: "Another Name"
> ```

```
class Person
  new: (name = nil) =>
    unless name
      @name = @getName!

  getName: => 'Default name'

class Another extends Person
  getName: 'Another default name'
```

I do not test it that all. ( writing from phone), but something like this should be work.

**Comment by Le0Developer** [comment]
You seem to have completely missed the point of this feature request, this request **is not** about how to do it.
It is about extending the existing special syntax `new: (@name) =>` to allow the use of `?` for the same reason we have the special syntax in the first place - compactness and avoid writing repetitive code.

---

```moon
class Person
    new: (@name) =>
```
is equivalent to
```moon
class Person
    new: (name) =>
        @name = name
```

And I want to request an extension of that special syntax to allow a `?`.

```moon
class Person
    new: (@name?) =>
```
should be equivalent to (as I've already explained in my first post)
```moon
class Person
    new: (name) =>
        @name = name if name?
```

**Comment by pigpigyyy** [comment]
IMO, the syntax `new: (@name) =>` is a bad design, and I prefer to write codes as
```moonscript
new: (name)=>
  @name = name
```
in a real project to increase readability and let others known what has been done in a clear way.

To implement this feature, we have to run extra patten matching rule for every argument that appears in the function declarations just for this corner use case. So I think maybe it's not worth implementing such seldomly used feature like this.

**Comment by Le0Developer** [comment]
It'd just be a "nice to have" to avoid having to write repetitive code.

Isn't that point of the existential operator, piping, macros and other new features aswell?

**Comment by pigpigyyy** [comment]
I found this feature won't slow down the parser and got it implemented by commit 14ac9a31f84c7bd25b37c4ae6ad133572a5df3a7. You may have a check for it.

**Comment by Le0Developer** [comment]
Thank you very much, looks good to me 😄
